torch/csrc/stable/library.h

```diff
@@ -76,7 +82,18 @@ template <
     typename T,
     std::enable_if_t<!detail::is_optional<T>::value, bool> = true>
 T to(StableIValue val) {
-  return *reinterpret_cast<T*>(&val);
+  static_assert(std::is_trivially_copyable_v<T>);
+  // T may not have a default constructor. (For example, it might be
+  // c10::Device.) However, std::memcpy implicitly creates a T at the
+  // destination. So, we can use a union to work around this lack of
+  // default constructor.
+  union Result {
```

Comment:
TIL about unions, clever

Reply:
https://dev-discuss.pytorch.org/t/unionizing-for-profit-how-to-exploit-the-power-of-unions-in-c/444

Reply:
Aren't this use of unions another UB?

Reply:
this is not "using unions for type punning". that phrasing typically refers to referencing a member of a union other than the active member.

the comment on lines 90-93 notes that std::memcpy implicitly starts the lifetime of a T at the destination. therefore, `t` is the active member of `Result` and it is legal to access.

also, that stack overflow discussion is about C, not C++.

Reply:
here's the citation for memcpy implicitly creating an object: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0593r6.html 
https://stackoverflow.com/questions/46840853/what-are-the-changes-if-any-to-the-memcpy-lifetime-initalization-rules-in-the claims that P0593R6 was adopted and retroactively applies to previous standard versions as well